openblas@0.3.20 %cce@13 build fails: ftn-2307 ftn: "-m" option must be followed by 0, 1, 2, 3 or 4

### Steps to reproduce the issue

`openblas@0.3.20 %cce@13.0.1` build fails using:
* `spack/develop` (2515cafb9c60851d5b218d1262212034f5606869 from `Wed Apr 27 16:08:33 2022 -0700`)
* `PrgEnv-cray/8.2.0`
* `cce/13.0.1`
* `SUSE Linux Enterprise Server 15 SP2`

Spack environment:
```
# spack/develop 2515cafb9c60851d5b218d1262212034f5606869
# Wed Apr 27 16:08:33 2022 -0700

spack:
  view: false
  concretization: separately
  config:
    concretizer: clingo
  compilers:
  - compiler:
      spec: cce@13.0.1
      paths:
        cc: cc
        cxx: CC
        f77: ftn
        fc: ftn
      operating_system: sles15
      target: any
      modules:
      - PrgEnv-cray
      - cce/13.0.1
  - compiler:
      spec: gcc@7.5.0
      paths:
        cc: /usr/bin/gcc
        cxx: /usr/bin/g++
        f77: /usr/bin/gfortran
        fc: /usr/bin/gfortran
      operating_system: sles15
      target: any
  packages:
    all:
      compiler: [cce@13.0.1]
  specs:
  - openblas@0.3.20 ~bignuma~consistent_fpcsr~ilp64+locking+pic+shared symbol_suffix=none threads=openmp %cce ^ncurses%gcc
```

Concretization:
```
 -   elprk2s  openblas@0.3.20%cce@13.0.1~bignuma~consistent_fpcsr~ilp64+locking+pic+shared symbol_suffix=none threads=openmp arch=cray-sles15-zen
 -   nylrh3j      ^perl@5.34.1%cce@13.0.1+cpanm+shared+threads arch=cray-sles15-zen
 -   f4matl7          ^berkeley-db@18.1.40%cce@13.0.1+cxx~docs+stl patches=b231fcc arch=cray-sles15-zen
 -   ughig26          ^bzip2@1.0.8%cce@13.0.1~debug~pic+shared arch=cray-sles15-zen
 -   xymlw7q              ^diffutils@3.8%cce@13.0.1 arch=cray-sles15-zen
 -   yjofnqb                  ^libiconv@1.16%cce@13.0.1 libs=shared,static arch=cray-sles15-zen
 -   ek5tefr          ^gdbm@1.19%cce@13.0.1 arch=cray-sles15-zen
 -   j77pugq              ^readline@8.1%cce@13.0.1 arch=cray-sles15-zen
 -   5peuggt                  ^ncurses@6.2%gcc@7.5.0~symlinks+termlib abi=none arch=cray-sles15-zen
 -   3javkah                      ^pkgconf@1.8.0%gcc@7.5.0 arch=cray-sles15-zen
 -   6ln5n2o          ^zlib@1.2.12%cce@13.0.1+optimize+pic+shared patches=0d38234 arch=cray-sles15-zen
```

Reproducer:
```
$> ssh perlmutter
eugene@perlmutter:login19:~> git clone https://github.com/spack/spack
eugene@perlmutter:login19:~> (cd spack && git checkout 2515cafb)
eugene@perlmutter:login19:~> export SPACK_DISABLE_LOCAL_CONFIG=1
eugene@perlmutter:login19:~> . spack/share/spack/setup-env.sh
eugene@perlmutter:login19:~> spack -e . install
...
==> Installing openblas-0.3.20-elprk2sqjmxmwfzkhdpoqvd7h2cgfuf6
==> No binary for openblas-0.3.20-elprk2sqjmxmwfzkhdpoqvd7h2cgfuf6 found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/84/8495c9affc536253648e942908e88e097f2ec7753ede55aca52e5dead3029e3c.tar.gz
==> No patches needed for openblas
==> openblas: Executing phase: 'edit'
==> openblas: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j12' 'CC=/global/cfs/cdirs/m3503/ParaTools/openblas/spack/lib/spack/env/cce/cc' 'FC=/global/cfs/cdirs/m3503/ParaTools/openblas/spack/lib/spack/env/cce/ftn' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'TARGET=ZEN' 'USE_LOCKING=1' 'USE_OPENMP=1' 'USE_THREAD=1' 'RANLIB=ranlib' 'libs' 'netlib' 'shared'

14 errors found in build log:
     2859    ftn-2307 ftn: ERROR in command line
     2860      The "-m" option must be followed by 0, 1, 2, 3 or 4.
     2861    ftn-2307 ftn: ERROR in command line
     2862      The "-m" option must be followed by 0, 1, 2, 3 or 4.
     2863    ftn-2307 ftn: ERROR in command line
     2864      The "-m" option must be followed by 0, 1, 2, 3 or 4.
  >> 2865    make[2]: *** [<builtin>: sgbrfs.o] Error 1
     2866    make[2]: *** Waiting for unfinished jobs....
  >> 2867    make[2]: *** [<builtin>: spotrf2.o] Error 1
  >> 2868    make[2]: *** [<builtin>: sgbcon.o] Error 1
  >> 2869    make[2]: *** [<builtin>: sgetrf2.o] Error 1
  >> 2870    make[2]: *** [<builtin>: sgbbrd.o] Error 1
  >> 2871    make[2]: *** [<builtin>: sgbequ.o] Error 1
  >> 2872    make[2]: *** [<builtin>: sgbtrf.o] Error 1
  >> 2873    make[2]: *** [<builtin>: sgbtf2.o] Error 1
  >> 2874    make[2]: *** [<builtin>: sgbsvx.o] Error 1
  >> 2875    make[2]: *** [<builtin>: sgbsv.o] Error 1
  >> 2876    make[2]: *** [<builtin>: sgebak.o] Error 1
  >> 2877    make[2]: *** [<builtin>: sgbtrs.o] Error 1
     2878    make[2]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-elprk2sqjmxmwfzkhdpoqvd7h2cgfuf6/spack-src/lapack-netlib/SRC'
  >> 2879    make[1]: *** [Makefile:27: lapacklib] Error 2
     2880    make[1]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-elprk2sqjmxmwfzkhdpoqvd7h2cgfuf6/spack-src/lapack-netlib'
  >> 2881    make: *** [Makefile:250: netlib] Error 2
```


### Error message

<details><summary>Error message</summary><pre>
==> Installing openblas-0.3.20-elprk2sqjmxmwfzkhdpoqvd7h2cgfuf6
==> No binary for openblas-0.3.20-elprk2sqjmxmwfzkhdpoqvd7h2cgfuf6 found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/84/8495c9affc536253648e942908e88e097f2ec7753ede55aca52e5dead3029e3c.tar.gz
==> No patches needed for openblas
==> openblas: Executing phase: 'edit'
==> openblas: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j12' 'CC=/global/cfs/cdirs/m3503/ParaTools/openblas/spack/lib/spack/env/cce/cc' 'FC=/global/cfs/cdirs/m3503/ParaTools/openblas/spack/lib/spack/env/cce/ftn' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'TARGET=ZEN' 'USE_LOCKING=1' 'USE_OPENMP=1' 'USE_THREAD=1' 'RANLIB=ranlib' 'libs' 'netlib' 'shared'

14 errors found in build log:
     2859    ftn-2307 ftn: ERROR in command line
     2860      The "-m" option must be followed by 0, 1, 2, 3 or 4.
     2861    ftn-2307 ftn: ERROR in command line
     2862      The "-m" option must be followed by 0, 1, 2, 3 or 4.
     2863    ftn-2307 ftn: ERROR in command line
     2864      The "-m" option must be followed by 0, 1, 2, 3 or 4.
  >> 2865    make[2]: *** [<builtin>: sgbrfs.o] Error 1
     2866    make[2]: *** Waiting for unfinished jobs....
  >> 2867    make[2]: *** [<builtin>: spotrf2.o] Error 1
  >> 2868    make[2]: *** [<builtin>: sgbcon.o] Error 1
  >> 2869    make[2]: *** [<builtin>: sgetrf2.o] Error 1
  >> 2870    make[2]: *** [<builtin>: sgbbrd.o] Error 1
  >> 2871    make[2]: *** [<builtin>: sgbequ.o] Error 1
  >> 2872    make[2]: *** [<builtin>: sgbtrf.o] Error 1
  >> 2873    make[2]: *** [<builtin>: sgbtf2.o] Error 1
  >> 2874    make[2]: *** [<builtin>: sgbsvx.o] Error 1
  >> 2875    make[2]: *** [<builtin>: sgbsv.o] Error 1
  >> 2876    make[2]: *** [<builtin>: sgebak.o] Error 1
  >> 2877    make[2]: *** [<builtin>: sgbtrs.o] Error 1
     2878    make[2]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-elprk2sqjmxmwfzkhdpoqvd7h2cgfuf6/spack-src/lapack-netlib/SRC'
  >> 2879    make[1]: *** [Makefile:27: lapacklib] Error 2
     2880    make[1]: Leaving directory '/tmp/lpeyrala/spack-stage/spack-stage-openblas-0.3.20-elprk2sqjmxmwfzkhdpoqvd7h2cgfuf6/spack-src/lapack-netlib'
  >> 2881    make: *** [Makefile:250: netlib] Error 2
</pre></details>


### Information on your system

* **Spack:** 0.18.0.dev0 (2515cafb9c60851d5b218d1262212034f5606869)
* **Python:** 3.6.15
* **Platform:** cray-sles15-zen3
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/8587159/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/8587161/spack-build-env.txt)

@prckent @haampie @siko1056 @lukebroskop @shahzebsiddiqui @wspear 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/openblas/cce.patch
+++ b/var/spack/repos/builtin/packages/openblas/cce.patch
@@ -1,0 +1,105 @@
+diff -ruN spack-src/cmake/fc.cmake spack-src-patched/cmake/fc.cmake
+--- spack-src/cmake/fc.cmake	2022-02-20 16:35:05.000000000 -0500
++++ spack-src-patched/cmake/fc.cmake	2022-08-04 13:02:13.294697807 -0400
+@@ -221,3 +221,14 @@
+   set(TIMER "NONE")
+ endif ()
+ 
++if (${F_COMPILER} STREQUAL "CRAY")
++  set(CCOMMON_OPT "${CCOMMON_OPT} -DF_INTERFACE_INTEL")
++  set(FCOMMON_OPT "${FCOMMON_OPT} -hnopattern")
++  if (INTERFACE64)
++    set (FCOMMON_OPT "${FCOMMON_OPT} -s integer64")
++  endif ()
++  if (NOT USE_OPENMP)
++    set(FCOMMON_OPT "${FCOMMON_OPT} -O noomp")
++  endif ()
++endif ()
++
+diff -ruN spack-src/cmake/system.cmake spack-src-patched/cmake/system.cmake
+--- spack-src/cmake/system.cmake	2022-02-20 16:35:05.000000000 -0500
++++ spack-src-patched/cmake/system.cmake	2022-08-04 13:03:15.173843680 -0400
+@@ -552,6 +552,14 @@
+   endforeach ()
+ endif ()
+ 
++if ("${F_COMPILER}" STREQUAL "NAG" OR "${F_COMPILER}" STREQUAL "CRAY")
++  set(FILTER_FLAGS "-msse3;-mssse3;-msse4.1;-mavx;-mavx2,-mskylake-avx512")
++  foreach (FILTER_FLAG ${FILTER_FLAGS})
++    string(REPLACE ${FILTER_FLAG} "" LAPACK_FFLAGS ${LAPACK_FFLAGS})
++    string(REPLACE ${FILTER_FLAG} "" LAPACK_FPFLAGS ${LAPACK_FPFLAGS})
++  endforeach ()
++endif ()
++
+ if ("${F_COMPILER}" STREQUAL "GFORTRAN")
+   # lapack-netlib is rife with uninitialized warnings -hpa
+   set(LAPACK_FFLAGS "${LAPACK_FFLAGS} -Wno-maybe-uninitialized")
+diff -ruN spack-src/f_check spack-src-patched/f_check
+--- spack-src/f_check	2022-02-20 16:35:05.000000000 -0500
++++ spack-src-patched/f_check	2022-08-04 13:05:03.075337778 -0400
+@@ -34,7 +34,7 @@
+ 	      "pathf90", "pathf95",
+ 	      "pgf95", "pgf90", "pgf77", "pgfortran", "nvfortran",
+ 	      "flang", "egfortran",
+-              "ifort", "nagfor");
++              "ifort", "nagfor", "ftn", "crayftn");
+ 
+ OUTER:
+     foreach $lists (@lists) {
+@@ -76,6 +76,11 @@
+ 	    $vendor = FUJITSU;
+ 	    $openmp = "-Kopenmp";
+ 
++        } elsif ($data =~ /Cray/) {
++
++            $vendor = CRAY;
++	    $openmp = "-fopenmp";
++
+ 	} elsif ($data =~ /GNU/ || $data =~ /GCC/ ) {
+ 
+             $data =~ s/\(+.*?\)+//g;
+@@ -306,6 +311,10 @@
+ if ( $vendor eq "NAG") {
+ 	    $link = `$compiler $openmp -dryrun ftest2.f 2>&1 && rm -f a.out a.exe`;
+     }
++
++if ( $vendor eq "CRAY") {
++            $link = `$compiler $openmp -hnopattern ftest2.f 2>&1 && rm -f a.out a.exe`;
++    }
+ $linker_L = "";
+ $linker_l = "";
+ $linker_a = "";
+diff -ruN spack-src/Makefile.system spack-src-patched/Makefile.system
+--- spack-src/Makefile.system	2022-02-20 16:35:05.000000000 -0500
++++ spack-src-patched/Makefile.system	2022-08-04 13:01:53.413702690 -0400
+@@ -1271,6 +1271,19 @@
+ endif
+ endif
+ 
++ifeq ($(F_COMPILER), CRAY)
++CCOMMON_OPT += -DF_INTERFACE_INTEL
++FCOMMON_OPT += -hnopattern
++ifdef INTERFACE64
++ifneq ($(INTERFACE64), 0)
++FCOMMON_OPT += -s integer64
++endif
++endif
++ifneq ($(USE_OPENMP), 1)
++FCOMMON_OPT += -O noomp
++endif
++endif
++
+ ifdef BINARY64
+ ifdef INTERFACE64
+ ifneq ($(INTERFACE64), 0)
+@@ -1551,6 +1564,10 @@
+ ifeq ($(F_COMPILER),NAG)
+ LAPACK_FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
+ endif
++ifeq ($(F_COMPILER),CRAY)
++LAPACK_FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
++FFLAGS := $(filter-out -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mskylake-avx512 ,$(FFLAGS))
++endif
+ 
+ LAPACK_CFLAGS = $(CFLAGS)
+ LAPACK_CFLAGS += -DHAVE_LAPACK_CONFIG_H

--- a/var/spack/repos/builtin/packages/openblas/package.py
+++ b/var/spack/repos/builtin/packages/openblas/package.py
@@ -78,6 +78,9 @@ class Openblas(MakefilePackage):
     provides("lapack@3.9.1:", when="@0.3.15:")
     provides("lapack@3.7.0", when="@0.2.20")
 
+    # https://github.com/xianyi/OpenBLAS/pull/3712
+    patch("cce.patch", when="@0.3.20 %cce")
+
     # https://github.com/spack/spack/issues/31732
     patch("f_check-oneapi.patch", when="@0.3.20 %oneapi")
 

